packages/ws/src/struct/WebSocketManager.ts

```diff
@@ -0,0 +1,195 @@
+import { Awaitable, range } from '../utils/utils';
+
+/**
+ * Represents a range of shard ids
+ */
+export interface ShardRange {
+	start: number;
+	end: number;
+}
+
+/**
+ * Session information for a given shard, used to resume a session
+ */
+export interface SessionInfo {
+	/**
+	 * Session id for this shard
+	 */
+	sessionId: string;
+	/**
+	 * The sequence number of the last message sent by the shard
+	 */
+	sequence: number;
+	/**
+	 * The id of the shard
+	 */
+	shardId: number;
+	/**
+	 * The total number of shards at the time of this shard identifying
+	 */
+	shardCount: number;
+}
+
+/**
+ * Options for the WebSocketManager
+ */
+export interface WebSocketManagerOptions {
+	/**
+	 * The total number of shards across all WebsocketManagers you intend to instantiate.
+	 * Use `null` to use Discord's recommended shard count
+	 */
+	shardCount: number | null;
+	/**
+	 * The IDs of the shards this WebSocketManager should manage.
+	 * Use `null` to simply spawn 0 through `shardCount - 1`
+	 * @example
+	 * const manager = new WebSocketManager({
+	 *   shardIds: [1, 3, 7], // spawns shard 1, 3, and 7, nothing else
+	 * });
+	 * @example
+	 * const manager = new WebSocketManager({
+	 *   shardIds: {
+	 *     start: 3,
+	 *     end: 6,
+	 *   }, // spawns shards 3, 4, 5, and 6
+	 * });
+	 */
+	shardIds: number[] | ShardRange | null;
+	/**
+	 * The REST instance to use for fetching gateway information
+	 */
+	rest: REST;
+	/**
+	 * The gateway version to use
+	 * @default '10'
+	 */
+	version: string;
+	/**
+	 * Function used to retrieve session information (and resume) for a given shard
+	 * @example
+	 * const manager = new WebSocketManager({
+	 *   async retrieveSessionInfo(shardId): Awaitable<SessionInfo | null> {
+	 *     // Fetch this info from redis or similar
+	 *     return { sessionId: string, sequence: number };
+	 *     // Return null if no information is found
+	 *   },
+	 * });
+	 */
+	retrieveSessionInfo: (shardId: number) => Awaitable<SessionInfo | null>;
+	/**
+	 * Function used to store session information for a given shard
+	 */
+	updateSessionInfo: (sessionInfo: SessionInfo) => Awaitable<void>;
+}
+
+export class WebSocketManager {
+	/**
+	 * The options being used by this manager
+	 */
+	public readonly options: WebSocketManagerOptions;
+
+	// eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+	#token: string | null = null;
+
+	/**
+	 * Internal cache for a GET /gateway/bot result
+	 */
+	private gatewayInformation: {
+		data: APIGatewayBotInfo;
+		expiresAt: number;
+	} | null = null;
+
+	private readonly shards = new Collection<number, WebSocketShard>();
+
+	public constructor(options?: Partial<WebSocketManagerOptions>) {
+		this.options = { ...DefaultWebSocketManagerOptions, ...options };
+	}
+
+	/**
+	 * Sets the token this manager should use
+	 */
+	public setToken(token: string) {
+		this.#token = token;
+		this.options.rest.setToken(token);
```

Comment:
REST's `setToken`? Why exactly? Where else?

Reply:
Nowhere. It's expected the user provides a REST session for that token :P

Reply:
Well.. right now all the options are all.. optional and we default `rest` to `new REST()`. Should we only be calling it in that case?

Reply:
@kyranet @vladfrangu bump on this since there's now a clear path (i.e. making the `rest` instance required):
https://github.com/discordjs/discord.js/blob/818c2f12c3afae3a4c5a7607dcc183a45c5b5961/packages/ws/src/struct/WebSocketManager.ts#L195-L196

Reply:
As per Vlad, we'll be making passing a REST instance required.